AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLOperation.swift

```diff
@@ -87,48 +127,64 @@ final public class AWSGraphQLOperation<R: Decodable>: GraphQLOperation<R> {
         }
-        let requestPayload: Data
+
         do {
-            requestPayload = try JSONSerialization.data(withJSONObject: queryDocument)
+            return .success(try JSONSerialization.data(withJSONObject: queryDocument))
         } catch {
-            dispatch(result: .failure(APIError.operationError("Failed to serialize query document",
-                                                              "fix the document or variables",
-                                                              error)))
-            finish()
-            return
+            return .failure(APIError.operationError(
+                "Failed to serialize query document",
+                "fix the document or variables",
+                error
+            ))
         }
+    }
 
-        // Create request
-        let urlRequest = GraphQLOperationRequestUtils.constructRequest(with: endpointConfig.baseURL,
-                                                                       requestPayload: requestPayload)
-
-        Task {
-            // Intercept request
-            var finalRequest = urlRequest
-            for interceptor in requestInterceptors {
-                do {
-                    finalRequest = try await interceptor.intercept(finalRequest)
-                } catch let error as APIError {
-                    dispatch(result: .failure(error))
-                    cancel()
-                } catch {
-                    dispatch(result: .failure(APIError.operationError("Failed to intercept request fully.",
-                                                                      "Something wrong with the interceptor",
-                                                                      error)))
-                    cancel()
-                }
-            }
+    private func getEndpointConfig(from request: GraphQLOperationRequest<R>) -> Result<AWSAPICategoryPluginConfiguration.EndpointConfig, APIError> {
+        do {
+            return .success(try pluginConfig.endpoints.getConfig(for: request.apiName, endpointType: .graphQL))
+        } catch let error as APIError {
+            return .failure(error)
 
-            if isCancelled {
-                finish()
-                return
+        } catch {
+            return .failure(APIError.unknown("Could not get endpoint configuration", "", nil))
+        }
+    }
+
+    private func getEndpointInterceptors(from request: GraphQLOperationRequest<R>) -> Result<AWSAPIEndpointInterceptors?, APIError> {
+        getEndpointConfig(from: request).flatMap { endpointConfig in
+            do {
+                if let pluginOptions = request.options.pluginOptions as? AWSPluginOptions,
+                   let authType = pluginOptions.authType
+                {
+                    return .success(try pluginConfig.interceptorsForEndpoint(
+                        withConfig: endpointConfig,
+                        authType: authType
+                    ))
```

Comment:
No action needed, just some context about this code. This is what provides the ability for DataStore to support models with multi-auth, by allowing DataStore to specific toggle between authType at runtime to attempt the synchronization calls.

When developers use API plugin directly, the configuration's authType is currently the only way to control what's in the interceptors. Developers have to duplicate the api sections in the configuration and change authType to control which whether default interceptor is in prelude/postlude, ie. a separate section for API_KEY, COGNITO_USER_POOL, with a key for each section, the key is then used as the apiName in the request object.

This is a huge customer pain point, especially Amplify CLI may also revert the changes to amplifyconfiguration.json. A runtime solution can potentially be to expose it differently as a public API or document the current mechanism. Solving this however will require some additional thoughts on customer defined interceptors. This code re-creates the interceptors based on authType, which creates a new struct, losing the customer's added interceptors since this is a mutating func on the existing config object. The other customer defined interceptors, such as the lambda or OIDC interceptors should continue to work as expected.